### PR TITLE
Proposal for RemoteWrite CRD

### PR DESCRIPTION
**Pull request by superbrothers**
This PR adds a proposal document for RemoteWrite CRD.

ref. https://github.com/prometheus-operator/prometheus-operator/issues/6508

cc @simonpasquier

### CODE REVIEW

File: Documentation/proposals/202407-remote-write.md
Comment (simonpasquier) [inline Documentation/proposals/202407-remote-write.md:99]: Side-note: the new CRD would need to be added to the list of supported resources.

File: Documentation/proposals/202407-remote-write.md
Comment (simonpasquier) [inline Documentation/proposals/202407-remote-write.md:105]: From I can see, it should work the same as the Prometheus CRD. Is there anything that would differ?

File: Documentation/proposals/202407-remote-write.md
Comment (simonpasquier) [inline Documentation/proposals/202407-remote-write.md:109]: > The generated configuration will be stored in a secret mounted in the Prometheus pod.

No change to the current way, right?

File: Documentation/proposals/202407-remote-write.md
Comment (simonpasquier) [inline Documentation/proposals/202407-remote-write.md:44]: It might be good to mention that secret/configmap refs will be resolved in the RemoteWrite namespace.

**Review by simonpasquier — COMMENTED** [review]
Thanks for starting this! My main question would be how admins can protect the Prometheus instance from "expensive" configurations (see also https://github.com/prometheus/prometheus/issues/13174). I don't have a good answer yet.

File: Documentation/proposals/202407-remote-write.md
Comment (philipgough) [inline Documentation/proposals/202407-remote-write.md:20]:
```suggestion
The main goal is to enable application developers/operators to self-service the remote write, and configure how the client sends metrics to the remote endpoint.
```

File: Documentation/proposals/202407-remote-write.md
Comment (superbrothers) [reply Documentation/proposals/202407-remote-write.md]: Fixed at 4d8afb2c6.

File: Documentation/proposals/202407-remote-write.md
Comment (superbrothers) [reply Documentation/proposals/202407-remote-write.md:93]: It seems to me that simply replacing the namespace labels does not prevent other tenants from stealing metrics. Should we only keep metrics for the namespace where the RemoteWrite object is created, as follows?

```yaml
write_relabel_configs:
...
- sourceLabels:
  - namespace
  regex: <namespace>
  action: keep
```

File: Documentation/proposals/202407-remote-write.md
Comment (superbrothers) [inline Documentation/proposals/202407-remote-write.md:104]:
```suggestion
```

File: Documentation/proposals/202407-remote-write.md
Comment (superbrothers) [reply Documentation/proposals/202407-remote-write.md:110]: Yes, it is still the same. The configuration file is gzipped, but essentially unchanged.

File: Documentation/proposals/202407-remote-write.md
Comment (superbrothers) [reply Documentation/proposals/202407-remote-write.md:44]:
```suggestion
    # Secret object references will be resolved in the namespace of this remoteWrite object.
    # In this manifest it is the "default" namespace.
    accessKey:
```

File: Documentation/proposals/202407-remote-write.md
Comment (superbrothers) [inline Documentation/proposals/202407-remote-write.md:48]:
```suggestion
    accessKey:
      # Secret object references will be resolved in the namespace of this remoteWrite object.
      # In this manifest it is the "default" namespace.
      name: aws_access
      key: access_key
```

File: Documentation/proposals/202407-remote-write.md
Comment (simonpasquier) [reply Documentation/proposals/202407-remote-write.md:108]: Thanks! Then it's probably simpler to say this at the beginning and that everything which is described for the Prometheus CRD will work the same for PrometheusAgent.

File: Documentation/proposals/202407-remote-write.md
Comment (simonpasquier) [reply Documentation/proposals/202407-remote-write.md:112]: Right so let's omit the things which won't change.
```suggestion
The Prometheus operator will generate the Prometheus configuration including remote_write configuration from the Prometheus CRD and the RemoteWrite resources matching remoteWriteSelector from the namespace(s) selected by remoteWriteNamespaceSelector for additional remote_rewrite configuration.
```

File: Documentation/proposals/202407-remote-write.md
Comment (superbrothers) [reply Documentation/proposals/202407-remote-write.md]: Updated in 29a6f2b21ea9c12efaf0ccf65b737b42995dd649.

File: Documentation/proposals/202407-remote-write.md
Comment (superbrothers) [reply Documentation/proposals/202407-remote-write.md]: Updated in 29b96b41a6f0ce1a3acd7d4efeac0a58eab22d40.

File: Documentation/proposals/202407-remote-write.md
Comment (ArthurSens) [inline Documentation/proposals/202407-remote-write.md:95]: Nice approach to respect tenants!

File: Documentation/proposals/202407-remote-write.md
Comment (ArthurSens) [inline Documentation/proposals/202407-remote-write.md:108]: This might sound as a silly question coming from a maintainer, but what does `excludedFromEnforcement` do again? 😅

**Review by ArthurSens — COMMENTED** [review]
One question:

Introducing a new CRD requires that the operator has RBAC permissions to read/list/watch this new resource. What happens if someone updates the operator to a version that has support for RemoteWrite CRD, but still hasn't configured the necessary permissions?

Could we do something similar to what we've done with the ScrapeConfig CRD? (allow starting the operator without permissions, do not reconcile if permissions aren't there)

File: Documentation/proposals/202407-remote-write.md
Comment (simonpasquier) [reply Documentation/proposals/202407-remote-write.md:108]: right now it's used to enforce a label (typically "namespace") with the "namespace of origin" value and it's supported for:
* metrics scraped from ServiceMonitor, PodMonitor, Probe objects
* rules from PrometheusRule objects

**Comment by simonpasquier** [comment]
+1 to the suggestion from @ArthurSens about checking the RBAC permissions of the operator. We could also add a feature gate but it's a bit overkill IMHO.

About protecting Prometheus against greedy remote write configs, we could allow the admins to define max values for the queue config settings (like queue capacity, sample age, max number of shards, ...).

File: Documentation/proposals/202407-remote-write.md
Comment (superbrothers) [reply Documentation/proposals/202407-remote-write.md:108]:
```suggestion
The RemoteWrite CRD should also respect the Prometheus `.spec.excludedFromEnforcement` field. It is a list of references to objects to be excluded from `enforcedNamespaceLabel`.
```

**Comment by superbrothers** [comment]
> Could we do something similar to what we've done with the ScrapeConfig CRD? (allow starting the operator without permissions, do not reconcile if permissions aren't there)

@ArthurSens  If it is possible with ScrapeConfig CRD, it should be possible here as well. I will try to achieve this in a similar way to ScrapeConfig.

**Comment by superbrothers** [comment]
> About protecting Prometheus against greedy remote write configs, we could allow the admins to define max values for the queue config settings (like queue capacity, sample age, max number of shards, ...).

@simonpasquier This can also be the case with other resources such as ServiceMonitor, where extreme configurations can affect the Prometheus server. It is difficult to support individual cases. I believe that admins should be responsible for protecting them, e.g. by using a ValidatingAdmissionPolicy.

**Comment by mviswanathsai** [comment]
> Thanks for starting this! My main question would be how admins can protect the Prometheus instance from "expensive" configurations (see also https://github.com/prometheus/prometheus/issues/13174). I don't have a good answer yet.

Not sure if this will help, but I wrote to the OP of that issue and learnt that they are writing everything to one/ a few endpoints and then configuring the endpoints to do the fragmentation. Not sure if we will even be able to enforce something like this. But this goes to show we can limit the number of RemoteWrite CRDs in the cluster if we really wanted to and not directly impact user's abilities. However, it definitely makes their life harder :\

**Comment by simonpasquier** [comment]
> > About protecting Prometheus against greedy remote write configs, we could allow the admins to define max values for the queue config settings (like queue capacity, sample age, max number of shards, ...).
> 
> @simonpasquier This can also be the case with other resources such as ServiceMonitor, where extreme configurations can affect the Prometheus server. It is difficult to support individual cases. I believe that admins should be responsible for protecting them, e.g. by using ValidatingAdmissionPolicy.


You can enforce scrape limits for ServiceMonitor. I don't want to dismiss your work here but we should document which options we have in the proposal :)

File: Documentation/proposals/202407-remote-write.md
Comment (mviswanathsai) [inline Documentation/proposals/202407-remote-write.md:106]: Pretty neat. But I am still not sure about how reliable this would be.

Comment (mviswanathsai) [reply Documentation/proposals/202407-remote-write.md:106]: Might be a nit, but for clarity, could we elaborate on what we mean by "foreign" metrics?

Comment (superbrothers) [reply Documentation/proposals/202407-remote-write.md:106]:
```suggestion
The item is inserted in the first position by the operator before any other user relabelings, so there's no way that users can have access to metrics labeled with namespaces other than the one in which the RemoteWrite object is created.
```

File: Documentation/proposals/202407-remote-write.md
Comment (superbrothers) [reply Documentation/proposals/202407-remote-write.md:26]:
```suggestion
Prometheus has two APIs for integrating with remote storage: 'remote write' and 'remote read'. Since there has been no confirmed demand to treat remote read as a CRD, we do not cover "RemoteRead" CRD here. We can still implement it later if needed.
```

**Comment by superbrothers** [comment]
@simonpasquier Do I need to make any other changes?

**Review by simonpasquier — COMMENTED** [review]
Thanks for the updated version! I'll get back to you this week hoperfully.
cc @philipgough who commented earlier :)

**Comment by dublx** [comment]
Nice, keeping an eye on the progress here, we have a need for this as well.

**Review by simonpasquier — COMMENTED** [review]
From my POV, it's ok to merge once you've the small suggested paragraph on admin limits.

File: Documentation/proposals/202407-remote-write.md
Comment (superbrothers) [reply Documentation/proposals/202407-remote-write.md:115]: Thanks!

File: Documentation/proposals/202407-remote-write.md
Comment (superbrothers) [inline Documentation/proposals/202407-remote-write.md:118]:
```suggestion
Similar to what exists for scrape resources (e.g. `enforcedSampleLimit` for `ServiceMonitor`), we will allow the Prometheus resource's owners to setup upper-bound limits on the remote write options. For example, it will be possible to disable metadata sending or limit the queue capacity:
```

**Review by simonpasquier — APPROVED** [review]
Thanks!

**Pull request merged**
